src/main/java/com/google/sps/data/Utility.java

```diff
@@ -61,179 +59,39 @@ public static String graphToJson(MutableGraph<GraphNode> graph, HashSet<String>
   /**
-   * Alternative function for calculating maxDepth
+   * Returns a mutable deep copy of the set of strings
    *
-   * @param graphInput the input graph, as a Mutatable Graph
-   * @param roots the name (string) of the roots
-   * @param graphNodesMap a mapping of strings to GraphNodes
-   * @param maxDepth the maximum depth of a node from a root
-   * @return a graph with nodes only a certain distance from a root
+   * @param roots the set of strings to deep copy
+   * @return a deep copy of the given set of strings
    */
-  public static MutableGraph<GraphNode> getGraphWithMaxDepth(
-      MutableGraph<GraphNode> graphInput,
-      Set<String> roots,
-      HashMap<String, GraphNode> graphNodesMap,
-      int maxDepth) {
-
-    MutableGraph<GraphNode> graphToReturn = GraphBuilder.directed().build();
-    if (maxDepth < 0) {
-      return graphToReturn; // If max depth below 0, then return an emtpy graph
-    }
-
-    Map<GraphNode, Boolean> visited = new HashMap<>();
-
-    for (String rootName : roots) {
-      GraphNode rootNode = graphNodesMap.get(rootName);
-      dfsVisit(rootNode, graphInput, visited, graphToReturn, maxDepth);
-    }
-    for (EndpointPair<GraphNode> edge : graphInput.edges()) {
-      if (graphToReturn.nodes().contains(edge.nodeU())
-          && graphToReturn.nodes().contains(edge.nodeV())) {
-        graphToReturn.putEdge(edge.nodeU(), edge.nodeV());
-      }
-    }
-
-    return graphToReturn;
+  public static HashSet<String> copyRoots(HashSet<String> roots) {
+    HashSet<String> copy = new HashSet<>();
+    copy.addAll(roots);
+    return copy;
   }
 
   /**
-   * Helper function for calculating max depth that actually visits a node and its children
+   * Returns a mutable deep copy of the given map from string to graph node
    *
-   * @param gn the GraphNode to visit
-   * @param graphInput the input graph
-   * @param visited a map that records whether nodes have been visited
-   * @param graphToReturn the graph to return, with all nodes within the specified depth
-   * @param depthRemaining the number of layers left to explore, decreases by one with each
-   *     recursive call on a child
+   * @param graphNodesMap the map from string to graph node to deepcopy
+   * @return a deep copy of the given map from string to Graph Node
    */
-  private static void dfsVisit(
-      GraphNode gn,
-      MutableGraph<GraphNode> graphInput,
-      Map<GraphNode, Boolean> visited,
-      MutableGraph<GraphNode> graphToReturn,
-      int depthRemaining) {
-    if (depthRemaining >= 0) {
-      visited.put(gn, true);
-      graphToReturn.addNode(gn);
-      for (GraphNode child : graphInput.successors(gn)) {
-        if (!visited.containsKey(child)) {
-          // Visit the child and indicate the increase in depth
-          dfsVisit(child, graphInput, visited, graphToReturn, depthRemaining - 1);
-        }
-      }
+  public static HashMap<String, GraphNode> copyNodeMap(HashMap<String, GraphNode> graphNodesMap) {
```

Comment:
Same thing as roots, maybe a generic copyHashMap function

Reply:
and maybe add tests for this as well

Reply:
Moved it back to DataGraph, I'll add tests to make sure it's a deep copy.